A43-sdk-authorization.md

```diff
@@ -0,0 +1,721 @@
+  	"name": {
+  		"description": "The name of an authorization policy. It is mainly for"
+  		  "monitoring and error message generation.",
+  		"type": "string"
+  	},
+  	"deny_rules": {
+  		"description": "List of deny rules to match. If a request matches any of the"
+  		  "deny rules, then it will be denied. If none of the deny rules matches or"
+  		  "there are no deny rules, the allow rules will be evaluated.",
+  		"type": "array",
+  		"items": {
+  			"$ref": "#/definitions/rule"
+  		}
+  	},
+  	"allow_rules": {
+  		"description": "List of allow rules to match. The allow rules will only be"
+  		  "evaluated after the deny rules. If a request matches any of the allow"
+  		  "rules, then it will allowed. If none of the allow rules matches, it will"
+  		  "be denied.",
+  		"type": "array",
+  		"items": {
+  			"$ref": "#/definitions/rule"
+  		}
+  	}
+  },
+  "required": ["name", "allow_rules"]
+}
+```
+
+#### Details
+
+gRPC SDK authorization policy has a list of *allow rules* and a list of *deny rules*.
+The following sequence is followed to make an authorization decision -
+1. Check for a match in *deny rules*, if a match is found, the request will be denied.
+   If no match is found in deny rules, or there are no deny rules execute next step.
+2. Check for a match in *allow rules*, if a match is found, the request will be allowed.
+   Note that allow rules is a required field here, if not present, policy is invalid.
+3. If no match is found in *allow rules*, we deny the request.
+
+Each *rule* has the following semantics -
+1. Each *rule* has a *name*, a *source* and a *request* to match against. For a *rule*
+   to match, both *source* and *request* must match.
+   - If both *source* and *request* are empty, the *rule* always matches. It is a
+     wildcard that can be used in a deny list to deny all RPCs or in an allow list to
+     accept all RPCs except those matching in the deny list.
+   - If only *source* is empty, we evaluate against the *request* fields and apply
+     that to any user (See example below). Similarly, if only *request* is empty, we
+     evaluate against the *source* fields and apply that to any action.
+2. Each *source* could contain a list of *principals*. The *principals* are ORed
+   together, i.e. it matches if one of them matches.
+   Sequence of steps to evaluate each principal from config -
+   1. If TLS is not used, matching fails.
+   2. If there is no client certificate, we get a match if principal is an empty
+      string.
+   3. If we have a client certificate, we check against certificate contents. A
+      match is found if principal matches URI SANs from the certificate. If
+      there are no URI SANs in certificate, or no match was found, we check
+      against DNS SANs. Similarly, if certificate has no DNS SANs or match wasn't
+      found, we check against Subject field from certificate.
+   
+   Consider the case where the *principals* list is empty, then we only check for
```

Comment:
In this case does principals have to be set to an empty list or is simply have the source object be present?

If an empty list is necessary, then that makes it incompatible with proto.

Reply:
Here, if the `principals: []`, then rule matches for any authenticated (TLS/mTLS) user, similar to `principals:["*"]`
Though the end result is the same, there are some differences.
In the "*" case, after the check for authenticated connection, since we first match against URI SANs, it will match against the first string and return successful match.

If it were an empty list (i.e. `principals: []`), then we can emulate the Envoy behavior of unset principal.
https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/rbac/v3/rbac.proto#config-rbac-v3-principal-authenticated
So, it will only check for authenticated connection, and return the match result accordingly.

For `source: {}` or missing source, principals would be set to ANY match.

Empty principals is implemented in Go, and C++ is in review.
https://github.com/grpc/grpc-go/blob/master/authz/rbac_translator.go#L165

Let me know your opinion. I can then accordingly update the proposal and related code.

Reply:
I see the current Envoy behavior as subtle and not something to emulate (we have to to a degree, but we don't have to be happy about it). I think we want this to remain proto-compatible, which means we can't special-case present-but-empty array.

You can get the same behavior via `principals: ["*", ""]`. So it seems these special semantics aren't necessary.

Reply:
Sure. Updated the proposal. Will make updates to the implementations.

Reply:
So missing source, `source: {}`, `principals: []` will all result in ANY rule type for Principal.

Reply:
@ejona86 I missed one of the reasons we chose empty principals. In TLS, where we do not have peer certificate, principals:["*"] would not match (won't match empty string since "*" is presence match). We may need to add special case in engine implementations in all 3 stacks to handle this.

Reply:
I thought I already mentioned that, and the solution:

> You can get the same behavior via `principals: ["*", ""]`. So it seems these special semantics aren't necessary.

Reply:
> But I felt it looked a bit hacky, since this representation is a result of internal implementation.

How so? `"*"` is defined as everything but empty string, so add an entry for empty string as well.

Reply:
Sure. Sounds good.